python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,361 @@
+                generated under (exclusive).
+            field_names: The partition key field names (i.e. column names for tabular
+                datasets). Required for HIVE partition paths, ignored for DIRECTORY
+                partition paths. When non-empty, the order and length of partition key
+                field names must match the order and length of partition values.
+        """
+        super(PathPartitionGenerator, self).__init__(style, base_dir, field_names)
+        if self.style == PartitionStyle.HIVE and not self.field_names:
+            raise ValueError(
+                "Hive partition path generation requires a corresponding list of "
+                "partition key field names. Please retry your request with one "
+                "or more field names specified."
+            )
+        generators = {
+            PartitionStyle.HIVE: self._generate_hive_partition_dirs,
+            PartitionStyle.DIRECTORY: self._generate_directory_partition_dirs,
+        }
+        self._generator_fn: Callable[[List[str]], List[str]] = generators.get(style)
+        if self._generator_fn is None:
+            raise ValueError(
+                f"Unsupported partition style: {style}. "
+                f"Supported styles: {generators.keys()}"
+            )
+
+    def _generate_hive_partition_dirs(self, values: List[str]) -> List[str]:
+        """Generates Hive directory names for the given values."""
+        return [f"{self._field_names[i]}={val}" for i, val in enumerate(values)]
+
+    def _generate_directory_partition_dirs(self, values: List[str]) -> List[str]:
+        """Generates Directory partition directory names for the given values."""
+        return values
+
+    def _generate_partition_dirs(self, values: List[str]) -> List[str]:
+        """Generates a list of partition directory names for the given values."""
+        return self._generator_fn(values)
+
+    def __call__(
+        self,
+        partition_values: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+    ) -> str:
+        """Returns the partition directory path for the given partition value strings.
+
+        All files for this partition should be written to this directory. If a base
+        directory is set, then the partition directory path will be given relative
+        to this base directory.
+
+        Args:
+            partition_values: The partition value strings to include in the partition
+                path. For HIVE partition paths, the order and length of partition
+                values must match the order and length of partition key field names.
+            filesystem: Filesystem that will be used for partition path file I/O.
+        """
+        self._normalize_base_dir(filesystem)
+        partition_dirs = self._generate_partition_dirs(partition_values)
+        return posixpath.join(self._normalized_base_dir, *partition_dirs)
+
+
+@DeveloperAPI
+class PathPartitionParser(PathPartitionBase):
```

Comment:
I think the tests added (which are nice) are against read apis. I wonder can we have unit test directly testing against those classes (the generator, parser etc)? It looks it should be relatively easy to test those interfaces.